text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,455 @@
+  stage limit that CodePipeline has. With the new implementation, the actions will automatically spread over multiple
+  stages, even if you deploy multiple applications in parallel.
+- For simple use cases, you no longer need to manage CodePipeline Artifacts: artifact management is implicit
+  as objects that produce artifacts can be used to reference them.
+
+
+### Examples
+
+The following:
+
+```ts
+import { SecretValue } from '@aws-cdk/core';
+import * as codepipeline from '@aws-cdk/aws-codepipeline';
+import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
+import * as cdkp from '@aws-cdk/pipelines';
+
+const sourceArtifact = new codepipeline.Artifact();
+const cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
+const integTestArtifact = new codepipeline.Artifact('IntegTests');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  cloudAssemblyArtifact,
+
+  // Where the source can be found
+  sourceAction: new codepipeline_actions.GitHubSourceAction({
+    actionName: 'GitHub',
+    output: sourceArtifact,
+    oauthToken: SecretValue.secretsManager('github-token'),
+    owner: 'OWNER'
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    commands: ['node ./integ-tests'],
+    additionalArtifacts: [integTestArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+import * as cdkp from '@aws-cdk/pipelines';
+
+const pipeline = new cdkp.Pipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: cdkp.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+  backend: new cdkp.AwsCodePipelineBackend(),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+pipeline.addCdkStage(stage, {
+  after: [
+    new cdkp.ShellScriptAction({
+      input: pipeline.build.additionalOutput('tests'),
+      commands: ['node ./integ-tests'],
+    }),
+  ],
+});
+```
+
+### Splitting definition and backend
+
+It is helpful if there is a clearly defined moment in time when the pipeline definition is ready to be rendered.
+
+There are various ways to organize this, each pretty much isomorphic to one another.
+
+OPTION 1
+
+Clear split of definition and instantiation. Very much like StepFunctions.
+
+```ts
+const pipelineDefinition = new cdkp.PipelineDefinition({
```

Comment:
This is a definition to distinguish it from the actual pipeline. 

I agree "definition" is not a great word. Blueprint?

Reply:
From a user perspective, this is the Pipeline (same as an `sns.Topic` is a topic). Our naming guidelines are pretty consistent in that sense - we name things based on what they represent in the real world, not based on their mechanical role.

Reply:
> From a user perspective, this is the Pipeline (same as an sns.Topic is a topic)

What I'm saying is that in this specific code example it's not, because this is proposing a split between the pipeline "form" and the pipeline instantiation by means of 2 different classes.

Reply:
I understand this option and I still think this should be called `Pipeline`. CDK pipeline is a concept that represents the deployment/rollout of the CDK app. From a user perspective, this object is the pipeline. The fact that you bind it to a backend later doesn't take away from that mental model.